daffodil-io/src/main/scala/org/apache/daffodil/io/DirectOrBufferedDataOutputStream.scala

```diff
@@ -33,30 +33,79 @@ import org.apache.daffodil.util.Maybe
 import org.apache.daffodil.util.Maybe._
 import org.apache.daffodil.util.MaybeULong
 import org.apache.daffodil.util.Misc
+import java.io.File
 
 /**
  * This simple extension just gives us a public method for access to the underlying byte array.
  * That way we don't have to make a copy just to access the bytes.
  */
-private[io] class ByteArrayOutputStreamWithGetBuf() extends java.io.ByteArrayOutputStream {
+private[io] class ByteArrayOrFileOutputStreamWithGetBuf(
+  chunkSizeInBytes: Long = 4096,
+  maxBufferSizeInBytes: Long = 2000 * (1 << 20))
+  extends java.io.ByteArrayOutputStream {
+  private var _byteArrayOS = this
+  private var _fileOS: java.io.FileOutputStream = null
+
+  var isFile: Boolean = false
+  var tempFile: File = null
+  var bytesWritten = count
+
+  def switchOS() = {
+    Assert.invariant(_fileOS == null)
+    tempFile = File.createTempFile("daffodil-", ".tmp")
+    _fileOS = new java.io.FileOutputStream(tempFile)
+    this.flush
+    this.writeTo(_fileOS)
+    bytesWritten = count
+    this.reset
+    isFile = true
+  }
+
+  override def write(b: Array[Byte]) = {
+    if ((_fileOS == null) && ((count + b.length) > maxBufferSizeInBytes))
```

Comment:
These ``_fileOS == null`` checks are kindof ugly. Thoughts on making this so this class just extends OutputStream instead of ByteArrayOutputStream, and then make it a wrapper for an output stream, which might change from ByteArrayOutputStream to FileOutputStream? I'm thinking something like:
```scala
  var stream = new ByteArrayOutputStream()

  @inline
  private def writeBytes(...): Unit {
    if (!isFile && needsSwitching) {
      val newStream = new java.io.FileOutputStream(...)
      stream.flush
      stream.asInstanceOf[ByteArrayOutputStream].writeTo(newStream)
      stream = newStream
    }
    bytesWritten += len
  }

  override def write(...) = {
    writeBytes(...)
    stream.write(...)
  }
}
```
Makes it so all the overrirde functions are basically just stream.whatever(), except for write which just calls the switch thing. Another benefit is once the switch happens, the old ByteArrayOutputStream can be garbage collected, whereas before it couldn't.

Reply:
I originally did extend OutputStream, but we were making heavy use of the count and buffer that ByteArrayOutputStream provides, but as long as we enforce the access of these things through the appropriate get() functions we should be able to just use OutputStream.